### ISSUE TITLE
GetFormatCharge() return 0 on complex ionic case smarts

### ISSUE THREAD

**Issue by rmrmg** [rdkit/rdkit#6526]
For non-trivial smarts, where charge can be easily determined (and has exactly one value for all possible variants) rdkit return 0 instead of expected value. Exempli gratia:
```
>>> m2=Chem.MolFromSmarts('[$([CX3-][CX4]):4]')
>>> m2.GetAtomWithIdx(0).GetFormalCharge()
0

>>> m2=Chem.MolFromSmarts('[CX3-,CX2-:4]')
>>> m2.GetAtomWithIdx(0).GetFormalCharge()
0
```
In both cases above I expected  -1.
Tested on version 2022.09.5 installed with conda on Linux (but probably it is very generic and not depends on OS/Python/conda)

**Comment by wopozka** [comment]
How would you define formal charge for a query? It works for normal mols, not query mols.

**Comment by rmrmg** [comment]
In smarts query atom is like [atom_defiition1 OR atom_definion2 OR atom_definion3, ...] 
e.g. [CX4H2,CX4H3], [CX3-,CX2-] or [CX3-,CX3+]
In all examples above you have defined two options (1st is CX4H2 or CX4H3 second is CX3- or CX2- last one is CX4- or CX3+).
When we talk about charges 1st smarts defined atom with format charge 0 (no matter if atom matches to CX4H2 or CX4H3 it needs to have no charge), second smarts define atom with -1 charge whereas third is +1 OR -1. Right now rdkit can correctly detect charge when there is no comma in atom definition. For example:
```
m = Chem.MolFromSmarts('[CX4-][CX3+]')
>>> m.GetAtomWithIdx(0).GetFormalCharge()
-1
>>> m.GetAtomWithIdx(1).GetFormalCharge()
1
```
In short, when it is possible to determine change, the true value should be return.

**Comment by wopozka** [comment]
Well I am not fully sure whether it should work for a defined lists of possibilities for given atom. In simple case, when there is a single definition I can agree. Having said that, IMHO there might be a bug included here as well. GetFormalCharge returns 0 as a charge, regardless whether it can or can't calculate anything eg:

```
print(Chem.MolFromSmarts('[CX3+0:1]').GetAtomWithIdx(0).GetFormalCharge())
0
```
 
```
print(Chem.MolFromSmarts('[CX3:1]').GetAtomWithIdx(0).GetFormalCharge())
0
```
In the first case it return charge 0, what is clearly defined in atom definition, but in the second case it returns also 0, while there is no charge definition at all, and thus it matches carbon with any charge.

**Comment by bp-kelley** [comment]
Objects created by smiles and smarts are not really the same thing.  While they share some of the same machinery, there is no implicit deduction in smarts like there is in a smiles string.  Note that this has to be true as smarts can define chemistry that RDKit sanitization rejects.

To understand what is going on, you need to use the "DescribeQuery" function on atoms and bonds:

```
>>> print(Chem.MolFromSmarts("[CX3+0:1]").GetAtomWithIdx(0).DescribeQuery())
AtomAnd
  AtomAnd
    AtomType 6 = val
    AtomTotalDegree 3 = val
  AtomFormalCharge 0 = val

>>> print(Chem.MolFromSmarts("[CX3:1]").GetAtomWithIdx(0).DescribeQuery())
AtomAnd
  AtomType 6 = val
  AtomTotalDegree 3 = val
```

Smiles tells you what an object is, Smarts tells you what it matches, these are not the same thing sometimes.  A real atom always has a charge, however in smarts, if there is no charge specified it is never checked so even though in practice the atom "says" the format charge is 0, it simply isn't used in that fashion.

**Comment by wopozka** [comment]
@bp-kelley Thanks for your explanation. The question is how is GetFormalCharge calculated for SMARTS. For proper, sanitized, mols everything is clear for me, but how it is done for SMARTS is not clear at all. I assumed (probably wrong) it is copied from AtomFormalCharge and then there should be a difference between AtomFormalCharge = 0 and lack of this parameter in query (I would assume in the first case GetFormalCharge should return 0, in the second probably None).

**Comment by bp-kelley** [comment]
Formal charge isn't calculated from smarts, it is converted into a matching tree as seen above.  Or more clearly, if you have an explicit format charge `[CX+1]` it will end up in the formal charge entry of the atom, but this is an implementation detail that you cannot count on.  If you don't specify it, or make it an OR or an AND, i.e. formal charge 1 or formal charge 2, it basically doesn't exist on the atom or is arbitrary.

So, when reading smarts, you can't really trust GetFormalCharge except under very specific circumstances.

**Comment by wopozka** [comment]
Therefore I've had (and still do) doubts whether we can say anything about formal charge in case of queries - imho not and we should not calculate it. But rdkit does it, and in this case if AtomFormalCharge is used for this purpose, then it should return AtomFormalCharge if all querry atom have the same value, or None in case nothing is set or AtomFormalCharge are different.
